kotlin-sdk-server/src/commonMain/kotlin/io/modelcontextprotocol/kotlin/sdk/server/ServerSession.kt

```diff
@@ -0,0 +1,371 @@
+import io.modelcontextprotocol.kotlin.sdk.Method.Defined
+import io.modelcontextprotocol.kotlin.sdk.PingRequest
+import io.modelcontextprotocol.kotlin.sdk.PromptListChangedNotification
+import io.modelcontextprotocol.kotlin.sdk.ResourceListChangedNotification
+import io.modelcontextprotocol.kotlin.sdk.ResourceUpdatedNotification
+import io.modelcontextprotocol.kotlin.sdk.SUPPORTED_PROTOCOL_VERSIONS
+import io.modelcontextprotocol.kotlin.sdk.ToolListChangedNotification
+import io.modelcontextprotocol.kotlin.sdk.shared.Protocol
+import io.modelcontextprotocol.kotlin.sdk.shared.RequestOptions
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.serialization.json.JsonObject
+
+private val logger = KotlinLogging.logger {}
+
+public open class ServerSession(private val serverInfo: Implementation, options: ServerOptions) : Protocol(options) {
+    @Suppress("ktlint:standard:backing-property-naming")
+    private var _onInitialized: (() -> Unit) = {}
+
+    @Suppress("ktlint:standard:backing-property-naming")
+    private var _onClose: () -> Unit = {}
+
+    init {
+        // Core protocol handlers
+        setRequestHandler<InitializeRequest>(Method.Defined.Initialize) { request, _ ->
+            handleInitialize(request)
+        }
+        setNotificationHandler<InitializedNotification>(Method.Defined.NotificationsInitialized) {
+            _onInitialized()
+            CompletableDeferred(Unit)
+        }
+    }
+
+    /**
+     * The capabilities supported by the server, related to the session.
+     */
+    private val serverCapabilities = options.capabilities
+
+    /**
+     * The client's reported capabilities after initialization.
+     */
+    public var clientCapabilities: ClientCapabilities? = null
+        private set
+
+    /**
+     * The client's version information after initialization.
+     */
+    public var clientVersion: Implementation? = null
+        private set
+
+    /**
+     * Registers a callback to be invoked when the server has completed initialization.
+     */
+    public fun onInitialized(block: () -> Unit) {
+        val old = _onInitialized
+        _onInitialized = {
+            old()
+            block()
+        }
+    }
+
+    /**
+     * Registers a callback to be invoked when the server session is closing.
+     */
+    public fun onClose(block: () -> Unit) {
+        val old = _onClose
+        _onClose = {
+            old()
+            block()
+        }
+    }
+
+    /**
+     * Called when the server session is closing.
+     */
+    override fun onClose() {
+        logger.debug { "Server connection closing" }
+        _onClose()
+    }
+
+    /**
+     * Sends a ping request to the client to check connectivity.
+     *
+     * @return The result of the ping request.
+     * @throws IllegalStateException If for some reason the method is not supported or the connection is closed.
+     */
+    public suspend fun ping(): EmptyRequestResult = request<EmptyRequestResult>(PingRequest())
```

Comment:
nit: all explicit type arguments can be removed
```suggestion
    public suspend fun ping(): EmptyRequestResult = request<>(PingRequest())
```